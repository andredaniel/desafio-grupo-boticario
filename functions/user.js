const axios = require('axios')

const randomStatus = () => {
  const key = Math.floor(Math.random() * 3)

  const statuses = ['APPROVED', 'DISAPPROVED', 'IN_VALIDATION']

  return statuses[key]
}

exports.handler = async function (event, context) {
  const baseUrl = 'https://5fffb468cb21e10017af7e99.mockapi.io/api'

  if (event.httpMethod === 'GET') {
    const { data } = await axios.get(
      `${baseUrl}/orders/?user_id=${event.queryStringParameters.user_id}`
    )

    return {
      statusCode: 200,
      body: JSON.stringify(data),
    }
  }

  if (event.httpMethod === 'POST') {
    const newOrder = JSON.parse(event.body)

    newOrder.cashbackAppliedPercentage = Math.ceil(Math.random() * 20)
    newOrder.cashbackReceivedValue =
      (newOrder.value / 100) * newOrder.cashbackAppliedPercentage
    newOrder.status = randomStatus()

    const { data } = await axios.post(`${baseUrl}/orders`, newOrder)

    return {
      statusCode: 200,
      body: JSON.stringify(data),
    }
  }
}
