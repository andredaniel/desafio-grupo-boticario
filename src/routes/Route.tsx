import axios from 'axios'
import React, { useEffect } from 'react'
import {
  Redirect,
  Route as ReactDOMRoute,
  RouteProps as ReactDOMRouteProps,
  useHistory,
} from 'react-router-dom'
import { toast } from 'react-toastify'
import { useUser } from '../contexts/user.context'
import { isAuthenticated } from '../hooks/auth'

interface RouteProps extends ReactDOMRouteProps {
  isPrivate?: boolean
  forceRender?: boolean
  component: React.ComponentType
}

const Route: React.FC<RouteProps> = (props) => {
  const { isPrivate = false, component: Component, ...rest } = props
  const authenticated = isAuthenticated()
  const history = useHistory()
  const { user, setOrders, setIsLoading } = useUser()
  useEffect(() => {
    if (user) {
      history.listen(() => {
        axios
          .get(`/.netlify/functions/user/?user_id=${user.id}`)
          .then(({ data }) => {
            setOrders(data)
            setIsLoading(false)
          })
          .catch((error) => {
            toast.error(error.message)
          })
      })
    }
  }, [history, setIsLoading, setOrders, user, user?.id])

  return (
    <ReactDOMRoute
      {...rest}
      render={({ location }) => {
        if (isPrivate === authenticated) {
          return <Component />
        }

        return (
          <Redirect
            to={{
              pathname: authenticated ? '/' : '/login',
              state: { location },
            }}
          />
        )
      }}
    />
  )
}

export default Route
