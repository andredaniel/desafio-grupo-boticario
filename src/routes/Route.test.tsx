import { render } from '@testing-library/react'
import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'
import { UserProvider } from '../contexts/user.context'
import Route from './Route'

describe('<Route />', () => {
  it('should render Route component properly', () => {
    const { container } = render(
      <UserProvider>
        <Router>
          <Route component={() => <div>Route</div>} />
        </Router>
      </UserProvider>
    )
    expect(container).toMatchSnapshot()
  })
})
