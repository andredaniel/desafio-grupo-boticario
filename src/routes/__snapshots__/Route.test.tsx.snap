// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Route /> should render Route component properly 1`] = `
<div>
  <div>
    Route
  </div>
</div>
`;
